File: algorithms/julia/quicksort.jl
function partition!(arr::Vector{Int64}, l::Int, r::Int)
    pivot, i = arr[r], l
    for j in l:r-1
        if arr[j] <= pivot
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
        end
    end
    arr[i], arr[r] = arr[r], arr[i]
    return i
end

function quicksort!(arr::Vector{Int64}, l::Int, r::Int)
    l >= r && return
    p = partition!(arr, l, r)
    quicksort!(arr, l, p - 1)
    quicksort!(arr, p + 1, r)
end

function main()
    input = split(read(stdin, String))
    n     = parse(Int, input[1])
    arr   = parse.(Int64, input[2:n+1])

    # JIT warmup
    w = copy(arr[1:min(10, n)])
    quicksort!(w, 1, length(w))

    start = time_ns()
    quicksort!(arr, 1, n)
    ms = (time_ns() - start) / 1e6

    println(join(arr, " "))
    println(stderr, ms)
end

main()

File: algorithms/rust/quicksort.rs
use std::io::{self, Read, Write, BufWriter};
use std::time::Instant;

fn partition(arr: &mut [i64], l: usize, r: usize) -> usize {
    let pivot = arr[r];
    let mut i = l;
    for j in l..r {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, r);
    i
}

fn quicksort(arr: &mut [i64], l: usize, r: usize) {
    if l >= r { return; }
    let p = partition(arr, l, r);
    if p > 0 { quicksort(arr, l, p - 1); }
    quicksort(arr, p + 1, r);
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_ascii_whitespace();

    let n: usize = iter.next().unwrap().parse().unwrap();
    let mut arr: Vec<i64> = (0..n)
        .map(|_| iter.next().unwrap().parse().unwrap())
        .collect();

    let start = Instant::now();
    if n > 1 { quicksort(&mut arr, 0, n - 1); }
    let ms = start.elapsed().as_secs_f64() * 1000.0;

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    for (i, v) in arr.iter().enumerate() {
        if i > 0 { write!(out, " ").unwrap(); }
        write!(out, "{}", v).unwrap();
    }
    writeln!(out).unwrap();
    eprintln!("{:.3}", ms);
}